Add custom transports

Jake-Shadle opened the pull request:
Adds a `custom-transport` feature which enables the new `transport` module and `Options::set_transport` to specify your own HTTP transport for data from the Native SDK to be sent to your Sentry service.

I added a minimal example on the `Options::set_transport` method to show how you can just use a closure as I didn't want to pull in an actual HTTP client crate, but I can flesh out the example to be more real if you want me too. I'm using `reqwest` in the project where I was testing that this custom transport feature works correctly, for example.

Resolves: #3

Jake-Shadle committed:
Checkpointing

Jake-Shadle committed:
Fixup implementation and add docs

daxpedda commented on sentry-contrib-native-sys/src/lib.rs:
```diff
@@ -96,10 +96,31 @@ pub enum UserConsent {
 }
 
+/// This represents an interface for user-defined transports.
+#[repr(C)]
+pub struct Transport([u8; 0]);
```
```suggestion
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct Transport([u8; 0]);
```

daxpedda commented on sentry-contrib-native-sys/src/lib.rs:
```diff
@@ -96,10 +96,31 @@ pub enum UserConsent {
     Revoked = 0,
 }
 
+/// This represents an interface for user-defined transports.
```
```suggestion
/// This represents an interface for user-defined transports.
///
/// Transports are responsible for sending envelopes to sentry and are the last
/// step in the event pipeline. A transport has the following hooks, all of which
/// take the user provided `state` as last parameter. The transport state needs
/// to be set with `sentry_transport_set_state` and typically holds handles and
/// other information that can be reused across requests.
///
/// * `send_func`: This function will take ownership of an envelope, and is
///   responsible for freeing it via `sentry_envelope_free`.
/// * `startup_func`: This hook will be called by sentry and instructs the
///   transport to initialize itself.
/// * `shutdown_func`: Instructs the transport to flush its queue and shut down.
///   This hook receives a millisecond-resolution `timeout` parameter and should
///   return `true` when the transport was flushed and shut down successfully.
///   In case of `false`, sentry will log an error, but continue with freeing the
///   transport.
/// * `free_func`: Frees the transports `state`. This hook might be called even
///   though `shudown_func` returned `false` previously.
///
/// The transport interface might be extended in the future with hooks to flush
/// its internal queue without shutting down, and to dump its internal queue to
/// disk in case of a hard crash.
```

Just copy the full documentation from `sentry-native` here.

daxpedda commented on sentry-contrib-native-sys/src/lib.rs:
```diff
@@ -96,10 +96,31 @@ pub enum UserConsent {
+/// The Envelope is an abstract type which represents a payload being sent to
+/// sentry. It can contain one or more items, typically an Event.
+/// See <https://develop.sentry.dev/sdk/envelopes/>
+#[repr(C)]
+pub struct Envelope([u8; 0]);
```
```suggestion
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct Envelope([u8; 0]);
```

Jake-Shadle committed:
Update sentry-contrib-native-sys/src/lib.rs

Co-authored-by: daxpedda <daxpedda@gmail.com>

Jake-Shadle committed:
Update sentry-contrib-native-sys/src/lib.rs

Co-authored-by: daxpedda <daxpedda@gmail.com>

Jake-Shadle committed:
Update sentry-contrib-native-sys/src/lib.rs

Co-authored-by: daxpedda <daxpedda@gmail.com>

daxpedda commented on sentry-contrib-native-sys/src/lib.rs:
```diff
@@ -265,6 +310,50 @@ extern "C" {
+    ///
+    /// The return value needs to be freed with `sentry_string_free()`.
+    #[link_name = "sentry_envelope_serialize"]
+    pub fn envelope_serialize(envelope: *const Envelope, size: &mut usize) -> *const c_char;
```
```suggestion
    pub fn envelope_serialize(envelope: *const Envelope, size: *mut usize) -> *const c_char;
```

I know, just trying to keep this as straightforward a conversion as possible, dealing with anything more rust-like over in the other crate.

Jake-Shadle committed:
Update sentry-contrib-native-sys/src/lib.rs

Co-authored-by: daxpedda <daxpedda@gmail.com>

daxpedda commented on sentry-contrib-native-sys/src/lib.rs:
```diff
@@ -265,6 +310,50 @@ extern "C" {
+
+    /// Creates a new transport with an initial `send_func`.
+    #[link_name = "sentry_transport_new"]
+    pub fn transport_new(send_fn: SendEnvelopeFunction) -> *mut Transport;
```
```suggestion
    pub fn transport_new(send_func: Option<SendEnvelopeFunction>) -> *mut Transport;
```

daxpedda commented on sentry-contrib-native-sys/src/lib.rs:
```diff
@@ -265,6 +310,50 @@ extern "C" {
+    #[link_name = "sentry_transport_set_free_func"]
+    pub fn transport_set_free_func(
+        transport: *mut Transport,
+        free_fn: extern "C" fn(state: *mut c_void),
```
```suggestion
        free_func: Option<extern "C" fn(state: *mut c_void)>,
```

daxpedda commented on sentry-contrib-native-sys/src/lib.rs:
```diff
@@ -265,6 +310,50 @@ extern "C" {
+
+    /// Sets the transport startup hook.
+    #[link_name = "sentry_transport_set_startup_func"]
+    pub fn transport_set_startup_hook(transport: *mut Transport, startup_fn: StartupFunction);
```
```suggestion
    pub fn transport_set_startup_func(transport: *mut Transport, startup_func: Option<StartupFunction>);
```

daxpedda commented on sentry-contrib-native-sys/src/lib.rs:
```diff
@@ -265,6 +310,50 @@ extern "C" {
+    /// `true` in case all the pending envelopes have been sent within the timeout,
+    /// or `false` if the timeout was hit.
+    #[link_name = "sentry_transport_set_shutdown_func"]
+    pub fn transport_set_shutdown_func(transport: *mut Transport, shutdown_fn: ShutdownFunction);
```
```suggestion
    pub fn transport_set_shutdown_func(transport: *mut Transport, shutdown_func: Option<ShutdownFunction>);
```

Jake-Shadle committed:
Update sentry-contrib-native-sys/src/lib.rs

Co-authored-by: daxpedda <daxpedda@gmail.com>

Jake-Shadle committed:
Update sentry-contrib-native-sys/src/lib.rs

Co-authored-by: daxpedda <daxpedda@gmail.com>

Jake-Shadle committed:
Update sentry-contrib-native-sys/src/lib.rs

Co-authored-by: daxpedda <daxpedda@gmail.com>

Jake-Shadle committed:
Update sentry-contrib-native-sys/src/lib.rs

Co-authored-by: daxpedda <daxpedda@gmail.com>

Jake-Shadle committed:
Fixup suggestions

daxpedda commented:
I'm still looking through it, but I would like to minimize the implementation if possible.
Mainly the whole multi-threading part, I would strongly prefer this to be implemented by the user, I'm of yet unsure if that's even possible, but I think I will have a clearer picture after I finish going through the code.

I'm sorry, I can see it was a bit of work to implement all of this!

Jake-Shadle commented:
The threading is easy to remove, I just basically copied how the built-in transports were implemented to keep the runtime behavior relatively consistent.

daxpedda commented:
> The threading is easy to remove

Go ahead with that please.

We could additionally provide a threaded implementation with `reqwest` over a `feature = "reqwest-transport"`, I would even go as far as to make it the default. Doesn't have to be this PR though.
What do you think?

Jake-Shadle commented:
I have found that most crates that have built-in support for HTTP clients run into problems fairly quickly, as many crates that interact with HTTP already have their preferred client to use, with the features (eg. rustls vs openssl, etc) that they want. As soon as a dependency pulls in its own HTTP client, that dependency has to use the same features, and keep at a compatible version as the users own HTTP crate, otherwise they have more features than they want, possibly multiple versions of the same crate(s), or 2 or more different HTTP implementations because their choice differs from this crates.

That being said, I think an example of using eg reqwest would be nice to do, as that would only be a dev-dependency and doesn't place any requirements on the user, just gives them something to go off/copy.

daxpedda commented:
I have encountered the same issues myself as well, my though was that it was bettern then relying on `curl` or `winhttp`, but I guess I can decide on that later when I get more user feedback. I'm assuming you are leaning toward leaving the defaults as they are?

In any case, as you said, let's leave it as an example for now.

daxpedda commented on src/transport.rs:
```diff
@@ -0,0 +1,467 @@
+        }
+
+        {
+            let private = "http://a0b1c2d3e4f5678910abcdeffedcba12@192.168.1.1/0123456";
```
On a side note, did you get application crashes to work for Sentry on-premise? For me it always errors out on minidump processing and the logs show that the symbolicator fails on an unknown field `name`.

Jake-Shadle replied:
We use the hosted service, not sure what the difference would be between it and on-premise, but didn't have an problems that I saw.

daxpedda replied:
Alright, thanks.

Jake-Shadle committed:
Remove transport worker impl

Jake-Shadle committed:
Ignore clippy lint

Jake-Shadle committed:
Cleanup

Jake-Shadle committed:
Add custom transport example

Jake-Shadle committed:
Add feature note in README

Jake-Shadle commented:
Ok, I've refactored the transport to remove the background thread and so forth, which is now up to the user to implement if they want to.

I've also added an example using async Reqwest with rustls to show how it can be used.

Jake-Shadle committed:
Fixup

Jake-Shadle committed:
Ignore deprecated crates in examples

Jake-Shadle committed:
Rustfmt

daxpedda commented:
I see we are stepping over each others toes here 😄.
Thanks for the good work, but let's keep this PR to only adding transport.

Jake-Shadle commented:
Ahh ok, sure thing, I just wanted this to pass CI cleanly but I guess since it's still failing on master for various reasons you'll allow me to ignore it. ;)

daxpedda commented:
Of course 😉.

daxpedda commented:
Btw, don't worry about rebasing, I will rebase this for you when I'm done.

daxpedda commented:
![image](https://user-images.githubusercontent.com/1645124/85540644-88921700-b617-11ea-8c51-6c326d211d83.png)

Wanna click on that checkbox for me? This should give me write access to your branch.

Jake-Shadle commented:
That option isn't showing on this page... Maybe I broke something when I converted it to and back from a draft or something? I can try closing and creating a new one.

Jake-Shadle commented:
Oh wait yah, I have to have that checked when I create the pull request.

Jake-Shadle commented:
Nope, that didn't seem to work. :-1:

daxpedda commented:
Would you mind giving me collaborator status in your repo?

Jake-Shadle commented:
Ugh nope, can't do that either since I'm not in the organization owner list, pinging @repi

daxpedda commented:
Man, this is complicated, if nothing works out I can just post PR's on your repo.

Jake-Shadle commented:
I can just rebase?

daxpedda commented:
Yeah, go ahead.

Jake-Shadle committed:
Fix audit ignoring

Jake-Shadle committed:
Run prettier

repi commented:
@daxpedda added you as collaborator to our fork repo

daxpedda commented on src/transport.rs:
```diff
@@ -0,0 +1,423 @@
+    {
+        let mut builder = http::Request::builder();
+
+        {
+            let headers = builder
+                .headers_mut()
+                .ok_or_else(|| "unable to mutate headers")?;
+            headers.insert(
+                "user-agent",
+                USER_AGENT
+                    .parse()
+                    .map_err(|_| "failed to parse user agent")?,
+            );
+            headers.insert(
+                "content-type",
+                ENVELOPE_MIME
+                    .parse()
+                    .map_err(|_| "failed to parse MIME type")?,
+            );
+            headers.insert(
+                "accept",
+                "*/*".parse().map_err(|_| "failed to parse accept")?,
+            );
+        }
```
As far as I can see all of this should be unable to fail.
In the rest of the crate I handled most of the failures that could stem from `sentry-native` as panics. For example invalid UTF-8 or invalid MessagePack.

Any particular reason we shouldn't let it panic here too? What do you think about this issue in general?

daxpedda replied:
Just to clarify, I also mean the other errors that you built below, specifically the code that I marked here is irrelevant, sorry about that.

Jake-Shadle replied:
This was definitely overkill, because yes, none of these should fail. I'll just do unwraps.

daxpedda replied:
I'm looking through the C implementation of all the other failures you took into consideration, will remove some of them when I commit.

Jake-Shadle committed:
Just unwrap, these shouldn't fail

Jake-Shadle committed:
Could only fail due to bad headers

Jake-Shadle committed:
Fixup docs

daxpedda committed:
Fix compile errors, formatting and move some constants to sys.

daxpedda commented:
By the way, I just noticed even if you build your own custom transport, you still have to use curl or WinHttp because crashpad is using it (didn't confirm yet).

daxpedda committed:
Chane `Transport` into a trait and merge it with `TransportWorker`.

- Remove requirement from `set_transport` to pass a `Box`.
- Ability for `Options` to be borrowed.
- Change `transport::startup` to accept `&Options` instead of `Dsn` and `debug: bool`.
- Make `Transport` methods take `&mut self` instead of `&self`.
- Use `ManuallyDrop` instead of `Box::into_raw`.

Jake-Shadle commented:
It's possible, I only checked that turning off default transports on Linux resulted in openssl not being linked, but it doesn't use crashpad .

daxpedda commented:
Right, now that Windows is also supporting breakpad, WinHttp could be skipped there too.
FYI, crashpad support is coming to Linux too: getsentry/sentry-native#307.

daxpedda committed:
Clean up rest of `transport`.

- Make `transport` available by default, only hide things that need new dependencies behind feature.
- Instead of `str` errors add new error type.
- Add empty default implementations for `Transport` allowing a simple closure for `send` to work too.
- Catch panics in FFI and make them abort program.
- Move functions for handling envelopes to methods of envelopes themselves.
- Add derives to all structures.
- Hide `Dsn` fields and add getter and setter methods.
- Fix example and test.
- Fix bug in `Options` that made it impossible to use custom transport.
- Re-export `transport` constants.
- Fix Windows and macOS build with `--no-default-features`
- Add examples and `custom-transport` to CI.

daxpedda committed:
Fix CI.

daxpedda committed:
Fix stable clippy.

daxpedda reviewed (approved):
I'm going to fix documentation and add more tests in master, thanks!
I changed quite a bit, feel free to drop ur opinion on all of this!

daxpedda commented on examples/custom-transport.rs:
```diff
@@ -0,0 +1,201 @@
+// we can implement our own transport for Sentry data so that we don't pull in
+// C dependencies (COUGH OPENSSL COUGH) that we don't want
+struct Transport {
+    /// we don't currently use custom certs or proxies, so we can just use the
+    /// same client that the rest of ark uses, if we do start doing that, we
+    /// would need to build the client based on the options during startup()
```
> so we can just use the same client that the rest of **ark** uses

What is "ark"?

Jake-Shadle replied:
Oops, I just copied the implementation I used in our internal project and accidentally left the comment, you can just delete it.

daxpedda replied:
Alright!
